### PR TITLE
Tsvparser

### PR DESCRIPTION
**Pull request by ElyesDer**
**Description**
This PR adds support for Confidence value retrieval from TSV output, simplifying the process of filtering the recognized text.
Confidence evaluation feature was much requested and I thought maybe I could simplify access to text, and it's confidence value altogether.

**Implementation**
- [x] Added new Output.CONFIDENCE format with default value equal to .None
- [x] Support multiple output format from class Output_Confidence_Format (Details below)
- [x] Support text regrouping from .TSV 
- [x] Confidence value of Output_Confidence_Format.LINE or Output_Confidence_Format.FLAT are calculated based on the average of each recognized word in the TSV
- [x] Grouping text uses `block_num`, `par_num` and `line_num`

**Details & Screenshots**
Using the test image located in _/tests/data/test.jpg_
test
**LINE** -- Returns a line-by-line recognized text and a confidence based on the average score of each element.
`{1: {'text': 'This is a lot of 12 point text to test the', 'conf': 95.81818181818181}, 2: {'text': 'ocr code and see if it works on all types', 'conf': 94.4}, 3: {'text': 'of file format.', 'conf': 95.33333333333333}, 4: {'text': 'The quick brown dog jumped over the', 'conf': 96.0}, 5: {'text': 'lazy fox. The quick brown dog jumped', 'conf': 96.0}, 6: {'text': 'over the lazy fox. The quick brown dog', 'conf': 96.0}, 7: {'text': 'jumped over the lazy fox. The quick', 'conf': 96.0}, 8: {'text': 'brown dog jumped over the lazy fox.', 'conf': 96.0}}`

**WORD** -- Returns each recognized word, and it's confidence score.
`{0: {1: [{'text': 'This', 'conf': 96}, {'text': 'is', 'conf': 96}, {'text': 'a', 'conf': 95}, {'text': 'lot', 'conf': 95}, {'text': 'of', 'conf': 96}, {'text': '12', 'conf': 96}, {'text': 'point', 'conf': 96}, {'text': 'text', 'conf': 96}, {'text': 'to', 'conf': 96}, {'text': 'test', 'conf': 96}, {'text': 'the', 'conf': 96}], 2: [{'text': 'ocr', 'conf': 94}, {'text': 'code', 'conf': 94}, {'text': 'and', 'conf': 95}, {'text': 'see', 'conf': 96}, {'text': 'if', 'conf': 94}, {'text': 'it', 'conf': 94}, {'text': 'works', 'conf': 96}, {'text': 'on', 'conf': 94}, {'text': 'all', 'conf': 92}, {'text': 'types', 'conf': 95}], 3: [{'text': 'of', 'conf': 96}, {'text': 'file', 'conf': 95}, {'text': 'format.', 'conf': 95}]}, 1: {1: [{'text': 'The', 'conf': 96}, {'text': 'quick', 'conf': 96}, {'text': 'brown', 'conf': 96}, {'text': 'dog', 'conf': 96}, {'text': 'jumped', 'conf': 96}, {'text': 'over', 'conf': 96}, {'text': 'the', 'conf': 96}], 2: [{'text': 'lazy', 'conf': 96}, {'text': 'fox.', 'conf': 96}, {'text': 'The', 'conf': 96}, {'text': 'quick', 'conf': 96}, {'text': 'brown', 'conf': 96}, {'text': 'dog', 'conf': 96}, {'text': 'jumped', 'conf': 96}], 3: [{'text': 'over', 'conf': 96}, {'text': 'the', 'conf': 96}, {'text': 'lazy', 'conf': 96}, {'text': 'fox.', 'conf': 96}, {'text': 'The', 'conf': 96}, {'text': 'quick', 'conf': 96}, {'text': 'brown', 'conf': 96}, {'text': 'dog', 'conf': 96}], 4: [{'text': 'jumped', 'conf': 96}, {'text': 'over', 'conf': 96}, {'text': 'the', 'conf': 96}, {'text': 'lazy', 'conf': 96}, {'text': 'fox.', 'conf': 96}, {'text': 'The', 'conf': 96}, {'text': 'quick', 'conf': 96}], 5: [{'text': 'brown', 'conf': 96}, {'text': 'dog', 'conf': 96}, {'text': 'jumped', 'conf': 96}, {'text': 'over', 'conf': 96}, {'text': 'the', 'conf': 96}, {'text': 'lazy', 'conf': 96}, {'text': 'fox.', 'conf': 96}]}}`

**FLAT** -- Returns the recognized text as a block and a confidence score based on the average score of each element.
`{'text': ' This is a lot of 12 point text to test the ocr code and see if it works on all types of file format. The quick brown dog jumped over the lazy fox. The quick brown dog jumped over the lazy fox. The quick brown dog jumped over the lazy fox. The quick brown dog jumped over the lazy fox.', 'conf': 95.69393939393939}`

### CODE REVIEW

File: pytesseract/pytesseract.py [hunk 1]
```diff
@@ -556,6 +581,71 @@ def image_to_osd(
...
+                else:
+                    data[block_num][par_num] = {}
+                    data[block_num][par_num][line_num] = [tup]
+            else:
+                data[block_num] = {}
+                data[block_num][par_num] = {}
+                data[block_num][par_num][line_num] = [tup]
+
+    for _, b in data.items():
+        for _, l in b.items():
```
Comment (q0w) [inline pytesseract/pytesseract.py:612]: you can use `.values()` to get only values

Comment (ElyesDer) [reply pytesseract/pytesseract.py:612]: Indeed, I will make the required changes.

File: pytesseract/pytesseract.py [hunk 2]
```diff
@@ -556,6 +581,71 @@ def image_to_osd(
...
+            groupped[k] = l
+            k += 1
+
+    if config == None or config == 'word':
```
Comment (q0w) [inline pytesseract/pytesseract.py:616]:
```python
is None
```

**Comment by stefan6419846** [comment]
Why does this PR include changes to the code style as well, specifically changing the docstrings from double-quoted to single-quoted? This should not be subject of such a PR in my opinion.

**Comment by int3l** [comment]
@stefan6419846 in general this doesn't matter, because pre-commit and the config for this particular project will reject it, as we can see from the CI. Most probably, it is something related to their local development environment.

The more problematic part is that this PR changes the default behavior of the main function (if I recall correctly from the quick glance over the changes). This might break existing user setups. I am not sure if we even have enough coverage for the main function stuff.

And speaking of code coverage - I think that such chunks of functionality should have at least some kind of unit testing alongside.
It will make maintenance and future changes easier.

**Comment by ElyesDer** [comment]
> Why does this PR include changes to the code style as well, specifically changing the docstrings from double-quoted to single-quoted? This should not be subject of such a PR in my opinion.

Mostly related to my env. configuration, but were all fixed in following commits and the configured CI scripts.

> The more problematic part is that this PR changes the default behavior of the main function (if I recall correctly from the quick glance over the changes). This might break existing user setups. I am not sure if we even have enough coverage for the main function stuff.

True, changes should not have made it to the main function, even though it has demonstration purpose only.

> And speaking of code coverage - I think that such chunks of functionality should have at least some kind of unit testing alongside.

Totally agree, this was a rushed addition, will add some unit testing scenarios on the roadmap.

**Pull request closed without merging**